src/ML/src/MANN.cpp

```diff
@@ -0,0 +1,385 @@
+        return false;
+    }
+
+    // resize the output
+    m_pimpl->structuredOutput.rawData.resize(outputSize);
+    m_pimpl->structuredOutput.shape[0] = 1; // batch
+    m_pimpl->structuredOutput.shape[1] = outputSize;
+
+    // create tensor required by onnx
+    m_pimpl->structuredOutput.tensor
+        = Ort::Value::CreateTensor<float>(m_pimpl->memoryInfo,
+                                          m_pimpl->structuredOutput.rawData.data(),
+                                          m_pimpl->structuredOutput.rawData.size(),
+                                          m_pimpl->structuredOutput.shape.data(),
+                                          m_pimpl->structuredOutput.shape.size());
+    // populate variable handler related to the output
+    // the serialization matters
+    m_pimpl->structuredOutput.handler.addVariable("future_base_positions",
+                                                  2 * projectedBaseHorizon / 2);
+    m_pimpl->structuredOutput.handler.addVariable("future_facing_directions",
+                                                  2 * projectedBaseHorizon / 2);
+    m_pimpl->structuredOutput.handler.addVariable("future_base_velocities",
+                                                  2 * projectedBaseHorizon / 2);
+    m_pimpl->structuredOutput.handler.addVariable("joint_positions", numberOfJoints);
+    m_pimpl->structuredOutput.handler.addVariable("joint_velocities", numberOfJoints);
+    m_pimpl->structuredOutput.handler.addVariable("base_velocity", 3);
+
+    // resize the output
+    m_pimpl->output.futureBasePositionTrajectory.resize(2, projectedBaseHorizon / 2);
+    m_pimpl->output.futureFacingDirectionTrajectory.resize(2, projectedBaseHorizon / 2);
+    m_pimpl->output.futureBaseVelocitiesTrajectory.resize(2, projectedBaseHorizon / 2);
+    m_pimpl->output.jointPositions.resize(numberOfJoints);
+    m_pimpl->output.jointVelocities.resize(numberOfJoints);
+    m_pimpl->output.projectedBaseVelocity = manif::SE2Tangentd::Zero();
+
+    return true;
+}
+
+bool MANN::setInput(const MANNInput& input)
+{
+    return m_pimpl->populateInput(input);
+}
+
+bool MANN::advance()
+{
+    auto unpackMatrix
+        = [this](const std::string& variableName, Eigen::Ref<Eigen::MatrixXd> matrix) {
+              const auto& variable = m_pimpl->structuredOutput.handler.getVariable(variableName);
+
+              assert(variable.isValid());
+
+              // the matrix has been already allocated
+              const std::size_t rows = matrix.rows();
+              const std::size_t cols = matrix.cols();
+
+              Eigen::Ref<const Eigen::VectorXd> mean
+                  = m_pimpl->structuredOutput.mean.segment(variable.offset, variable.size);
+              Eigen::Ref<const Eigen::VectorXd> std
+                  = m_pimpl->structuredOutput.std.segment(variable.offset, variable.size);
+
+              // TODO try to avoid to do memory allocation
+              Eigen::VectorXd temp
```

Comment:
You could have a map with key the name (e.g. "future_base_positions") and as value a buffer. The call to this map can be hidden in the labda (since you are already capturing ``this``). You can avoid to initialize it, but at least it will survive after the first iteration.

Reply:
After https://github.com/ami-iit/mann-pytorch/pull/5 I cleaned the code and temp is not required anymore :)